crates/provider/src/layers/join_fill.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{PendingTransactionBuilder, Provider, ProviderLayer, RootProvider};
+use alloy_network::{Ethereum, Network};
+use alloy_transport::{Transport, TransportResult};
+use async_trait::async_trait;
+use futures::try_join;
+use std::{future::Future, marker::PhantomData};
+
+/// A layer that can fill in a `TransactionRequest` with additional information.
+///
+/// ## Lifecycle Notes
+///
+/// - `ready` MUST be called before `request` and `fill`. It is acceptable to panic in `request` and
+///   `fill` if `ready` has not been called.
+/// - the output of `request` MUST be passed to `fill` before `finished()`
+/// is called again.
+pub trait TxFiller<N: Network = Ethereum>: Clone + Send + Sync {
+    /// The properties that this filler retrieves from the RPC. to fill in the
+    /// TransactionRequest.
+    type Fillable: Send + Sync + 'static;
+
+    /// Joins this filler with another filler to compose multiple fillers.
+    fn join_with<T>(self, other: T) -> JoinFill<Self, T, N>
+    where
+        T: TxFiller<N>,
+    {
+        JoinFill::new(self, other)
+    }
+
+    /// Returns true if the filler is ready to fill in the transaction request.
```

Comment:
when would this not be the case?
if from is none for estimation for example?

Reply:
if `from` is none for `nonce`, that's a big one